src/2015-04-25-epidemiological-compartment-model.cpp

```diff
@@ -0,0 +1,240 @@
+ *
+ * This post uses a classic 
+ * Susceptible-Infected-Recovered (SIR)
+ * [epidemiological compartment model](https://en.wikipedia.org/wiki/Compartmental_models_in_epidemiology).
+ * Compartment models are simple, commonly-used dynamical
+ * systems models.  Here I demonstrate the
+ * [tau-leap method](https://en.wikipedia.org/wiki/Tau-leaping), where
+ * a discrete number of individuals move probabilistically 
+ * between compartments at fixed intervals in time.
+ * In this model, the wait-times within class 
+ * are exponentially distributed, and the number of
+ * transitions between states in a fixed time step are 
+ * Poisson distributed.
+ *
+ * This model is parameterized for the spread of measles
+ * in a closed population, 
+ * where the birth rate (nu) = death rate (mu).
+ * The transmission rate (beta) describes how frequently 
+ * susceptible (S) and infected (I) individuals come into 
+ * contact, and the recovery rate (gamma) describes the
+ * the average time an individual spends infected before
+ * recovering. 
+*/
+
+/** 
+ * ### C++ Code
+ * Note that cpp Functions must be marked with the following comment like for use in R
+ * // [[Rcpp::export]]
+*/
+
+#include <Rcpp.h>
+// Declare commonly used objects
+using Rcpp::NumericVector;
+using Rcpp::List;
+using Rcpp::DataFrame;
+using Rcpp::Named;
+// use Rcpp namespace R:: (alias to R API)
+// R::rpois always returns a single value
+// to return multiple (e.g. Integer/NumericVector, 
+// use Rcpp::rpois(int ndraw, param) and friends
+using R::rpois;
+// rcpp also has a pmin for vectors
+using std::min;
+
+// This function will be used in R!
+// Evaluates the number of events 
+// and updates the states at each time step tau
+// [[Rcpp::export]]
+List tauleapCpp(List params){
+    // chained operations are tricky in cpp
+    // pull out list w/in list into its own object
+    List init = params["init"];
+    // need to "cast" R vector to cpp scalar
+    int nsteps = Rcpp::as<int>(params["nsteps"]);
+
+    // initialize each state vector in its own vector
+    // set all vals to initial vals
+    NumericVector SS(nsteps, init["S"]);
+    NumericVector II(nsteps, init["I"]);
+    NumericVector RR(nsteps, init["R"]);
+    NumericVector NN(nsteps, init["pop"]);
+    // fill time w/zeros
+    NumericVector time(nsteps);
+
+    // pull out params for easy reading 
+    double nu = params["nu"];
+    double mu = params["mu"];
+    double beta = params["beta"];
+    double gamma = params["gamma"];
+    double tau = params["tau"];
+
+    // Rcpp magic: Initialize RNG (uses R engine)
+    Rcpp::RNGScope scope;
+    // Calculate the number of events for each day,
+    // update state vectors
+    for (int istep = 0; istep < (nsteps-1); istep++){
+        // for easier reading
+        // use references to *today's* state values
+        // these should really be ints, 
+        // but rpois returns double, and I'm being lazy
+        double & iS = SS[istep];
```

Comment:
I think these should just be plain values rather than references. Reasons:

1. I could be wrong but it appears as if we never actually write back to the values so don't need the reference.

2. On a 64-bit system these references are the same exact size as the underlying double so don't provide any additional space efficiency.

3. The extra bit of `&` syntax might be confusing for users new to C++.